### PR TITLE
Fixes #9810 - IPMI facter extensions multichannel support

### PR DESCRIPTION
**Pull request by lzap**
This patch now fixes ipmi_facts and also add ipmi_N_facts (where N is channel
number) for those who would like to get information from multiple channels for
any reason.

Example run of facter with this patch on IBM server with IPMI on channel 1:

```
[root@ibm-x3655-01 ~]# FACTERLIB=/root facter | grep ipmi
ipmi_1_gateway => 10.16.71.254
ipmi_1_ipaddress => 10.16.68.91
ipmi_1_ipaddress_source => DHCP Address
ipmi_1_macaddress => 00:14:5e:5b:8c:8e
ipmi_1_subnet_mask => 255.255.248.0
ipmi_enabled => true
ipmi_gateway => 10.16.71.254
ipmi_ipaddress => 10.16.68.91
ipmi_ipaddress_source => DHCP Address
ipmi_macaddress => 00:14:5e:5b:8c:8e
ipmi_subnet_mask => 255.255.248.0
```

The same on HP box with IPMI on channel 2:

```
[root@hp-dl165g6-01 ~]# FACTERLIB=/root facter | grep ipmi
ipmi_2_gateway => 10.16.71.254
ipmi_2_ipaddress => 10.16.66.131
ipmi_2_ipaddress_source => DHCP Address
ipmi_2_macaddress => 00:25:b3:e6:5d:98
ipmi_2_subnet_mask => 255.255.248.0
ipmi_enabled => true
ipmi_gateway => 10.16.71.254
ipmi_ipaddress => 10.16.66.131
ipmi_ipaddress_source => DHCP Address
ipmi_macaddress => 00:25:b3:e6:5d:98
ipmi_subnet_mask => 255.255.248.0
```

@mhulan can you take a look?

### CODE REVIEW

**Comment by lzap** [comment]
@ares can you take a look please?

File: root/usr/share/fdi/facts/discovery-facts.rb [hunk 1]
```diff
@@ -109,34 +109,52 @@ def discovery_bootif
...
+        attributes[:subnet_mask] = $1
+      when /^MAC Address\s+: (.*)/
+        attributes[:macaddress] = $1
+      when /^Default Gateway IP\s+: (.*)/
+        attributes[:gateway] = $1
+      end
+    end if output
+
+    if attributes.keys.empty?
+      Facter.debug("Running ipmitool didn't give any information")
```
Comment (ares) [inline root/usr/share/fdi/facts/discovery-facts.rb]: maybe update the message so we know for what channel we have no info

File: root/usr/share/fdi/facts/discovery-facts.rb [hunk 2]
```diff
@@ -109,34 +109,52 @@ def discovery_bootif
...
+    end
+    unless default_found
+      attributes.each do |fact, value|
+        Facter.add("ipmi_#{fact}") do
```
Comment (ares) [inline root/usr/share/fdi/facts/discovery-facts.rb]: maybe extracting the Facter.add code to separate method would be better, it seems that only name and value differs

**Comment by ares** [comment]
I didn't tested but except small nits it looks good to me. I'm not IPMI expert so I don't know whether it's reasonable to consider IPMI with low channels as default but I suppose it does not do any harm.

**Comment by lzap** [comment]
Thanks I will fix these and release 2.2.

File: root/usr/share/fdi/facts/discovery-facts.rb (hunk 1 above)
Comment (lzap) [inline root/usr/share/fdi/facts/discovery-facts.rb]: Yeah, actually I've changed it and only report when the first channel was not yet found.

File: root/usr/share/fdi/facts/discovery-facts.rb (hunk 2 above)
Comment (lzap) [inline root/usr/share/fdi/facts/discovery-facts.rb]: Done.

**Comment by lzap** [comment]
Ok refactored a bit. Tested:

```
FACTERLIB=/root facter -d | grep ipmi
Running ipmitool on port 0 didn't give any information
Could not retrieve discovery_version: No such file or directory - /usr/share/fdi/VERSION
Could not retrieve discovery_release: No such file or directory - /usr/share/fdi/RELEASE
ipmi_1_gateway => 10.16.71.254
ipmi_1_ipaddress => 10.16.68.47
ipmi_1_ipaddress_source => DHCP Address
ipmi_1_macaddress => 00:1a:64:cc:36:e1
ipmi_1_subnet_mask => 255.255.248.0
ipmi_enabled => true
ipmi_gateway => 10.16.71.254
ipmi_ipaddress => 10.16.68.47
ipmi_ipaddress_source => DHCP Address
ipmi_macaddress => 00:1a:64:cc:36:e1
ipmi_subnet_mask => 255.255.248.0
```

Ignore the two errors, not related (testing it on non-discovery machine). @orrabin please merge at will.

**Comment by ohadlevy** [comment]
I have a vague memory that we used these facts somewhere, maybe it was staypuft, @ares maybe remembers?

**Comment by ares** [comment]
yes, these ipmi facts are used in staypuft and foreman has builtin support for them so similarly as it creates other interfaces, it can create bmc interface based on it

**Comment by ohadlevy** [comment]
so does it mean we need another patch for foreman?

**Comment by ares** [comment]
I don't think so, my understanding is that ipmi_* are always set to first channel we found, ipmi_n_* are just additional if someone would need to read info from more channels. AFAIK Foreman currently support only one bmc interface per host.

**Comment by lzap** [comment]
@ohadlevy to give you some more context - this was created for Staypuft team, there is a downstream BZ link. They read it in their codebase. The only issue is that some HP boxes have IPMI on channel 2 instead of 1. This patch fixes it and reports correctly in `ipmi_` facts (adding those special `ipmi_N`) just for case.

**Comment by lzap** [comment]
@stbenjam or @orrabin any chance you can take a look? I'd like to merge this and backport this downstream. Important patch, to test this you either need to boot image on hardware with BMC on 2nd channel, or to copy this file and use facter on a different (non-image) machine. I can provide you a hardware box if needed.

**Comment by lzap** [comment]
To test this: build and run discovery on IPMI enabled hardware. If you don't have any, I can provide you a shell on such a box and testing can be done manually running facter with this extension. I'd like to have this in the new fdi release. Thanks!

**Comment by stbenjam** [comment]
@lzap Yea, I'll need a hardware box with IPMI hardware to test this, if you can give me access to one.

**Comment by fernandocarolo** [comment]
FWIW, I have a setup with several HP boxes that report IPMI on channel 2 and this patch has worked fine with them. It also doesn't seem to break anything on other machines that have a more traditional config with IPMI on channel 1.

**Comment by lzap** [comment]
Have you extended the beaker machine I reserved to you? @stbenjam ? :-)

**Comment by stbenjam** [comment]
I know nothing about this... what beaker machine?

**Comment by lzap** [comment]
> I know nothing about this... what beaker machine?

Hmm sent you an email with reservation.

-- 
Later,
 Lukas #lzap Zapletal

**Comment by stbenjam** [comment]
I only have the e-mail from a few minutes ago, but anyway I'll take a look when I have time

**Comment by stbenjam** [comment]
Tested and looks good. Merged as 4608b2271c3f65273cc9e2ba28708ee6bcba05be, thanks @lzap!

**Pull request closed without merging**